pkg/client/client.go

```diff
@@ -113,7 +113,11 @@ func (c ApplyClient) applyObject(ctx context.Context, obj client.Object, options
 	if config.saveConfiguration {
 		// set current object as annotation
 		annotations := obj.GetAnnotations()
-		newConfiguration = getNewConfiguration(obj)
+		// reset the previous config to avoid recursive embedding of the object
+		if _, found := obj.GetAnnotations()[LastAppliedConfigurationAnnotationKey]; found {
+			delete(obj.GetAnnotations(), LastAppliedConfigurationAnnotationKey)
```

Comment:
I would rather move this to the `getNewConfiguration()` function. So the function always returns a clean object to be used in the last applied configuration annotation. And also would create a deep copy of the object in that `getNewConfiguration()` function before modifying it (removing the annotation). It won't break anything in the current code if we don't do it but it would make the function less error prone in case we modify the code in the future.

Reply:
Good point, done in https://github.com/codeready-toolchain/toolchain-common/pull/385/commits/aacc2c6479568c7e2dfba7bef4749098ea3336f6 

I've also added the annotation to the existing obj in unit tests.